### PR TITLE
Add api generator script with testing the api calls on mapactivity

### LINKED ISSUE #31: Kotlin API client generator

### PR DESCRIPTION
**Pull request by avan1235**
resolves #31 
generated files doesn't have ending line (will be fixed in future by setuping an action on github for formatting the code)

### CODE REVIEW

File: android/api_generate_kotlin.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+if [[ $# -ne 1 ]]; then
+    echo "Illegal number of parameters. Usage: ./api_generate_kotlin.sh api.yaml"
```
Comment (kantoniak) [inline android/api_generate_kotlin.sh]: :+1: for API path parameter

File: android/api_generate_kotlin.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+java -jar openapi-generator-cli.jar generate \
+     -g kotlin \
+     -i $1 \
+     -o api \
+     -p groupId='labs.pooh' -p modelMutable='false' -p packageName='labs.pooh.client' -p serializationLibrary='gson' -p dateLibrary='threetenbp' &&
+
+rm openapi-generator-cli.jar &&
+rm -rf app/src/main/java/labs/pooh/eaterslab/client &&
+cp -r api/src/main/kotlin/labs/pooh/client app/src/main/java/labs/pooh/eaterslab/ &&
+rm -rf ./api &&
```
Comment (kantoniak) [inline android/api_generate_kotlin.sh]: `mv` instead?

File: android/app/src/main/java/labs/pooh/eaterslab/MapSearchActivity.kt [hunk 3]
```diff
@@ -44,9 +50,18 @@ class MapSearchActivity : AbstractRevealedActivity() {
...
+
+    /**
+     * Resources that should be disposed when the fragment is destroyed
+     */
+    private val compositeDisposable = CompositeDisposable()
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
 
+        API = CafeteriaApi(getString(R.string.api_url))
```
Comment (kantoniak) [inline android/app/src/main/java/labs/pooh/eaterslab/MapSearchActivity.kt]: I think API url should come from gradle, not from static strings. This way you can build debug with let's say `localhost` and target release builds against official API endpoint. To discuss.

File: android/app/src/main/java/labs/pooh/eaterslab/ui/slideshow/SlideshowFragment.kt [hunk 4]
```diff
@@ -48,7 +48,7 @@ class SlideshowFragment : Fragment() {
         textSlideshow.text = getString(R.string.loading)
 
         val observables = toObservableBitmaps(imagesURLs.entries.toObservable())
-        val disposable = toDisposableSubscibe(observables, view)
+        val disposable = toDisposableSubscribe(observables, view)
```
Comment (kantoniak) [inline android/app/src/main/java/labs/pooh/eaterslab/ui/slideshow/SlideshowFragment.kt]: :+1:

File: android/app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -1,5 +1,6 @@
 <resources>
     <string name="app_name">EatersLab</string>
+    <string name="api_url">https://eaterslab.herokuapp.com/api</string>
```
Comment (kantoniak) [inline android/app/src/main/res/values/strings.xml]: As mentioned previously, API is config rather than translatable string.

Comment (kantoniak) [reply android/app/src/main/res/values/strings.xml]: Let's keep discussion in the previous thread.

File: android/gradle.properties [hunk 6]
```diff
@@ -6,7 +6,7 @@
 # http://www.gradle.org/docs/current/userguide/build_environment.html
 # Specifies the JVM arguments used for the daemon process.
 # The setting is particularly useful for tweaking memory settings.
-org.gradle.jvmargs=-Xmx256m -XX:MaxPermSize=256m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
+org.gradle.jvmargs=-Xmx320m -XX:MaxPermSize=320m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
```
Comment (kantoniak) [inline android/gradle.properties]: This config could go to ignored local gradle settings.

File: android/app/src/main/java/labs/pooh/eaterslab/client/apis/CafeteriaApi.kt [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+                throw ClientException(localVarError.body as? String ?: "Client error", localVarError.statusCode)
+            }
+            ResponseType.ServerError -> {
+                val localVarError = localVarResponse as ServerError<*>
+                throw ServerException(localVarError.message ?: "Server error", localVarError.statusCode)
+            }
+        }
+    }
+
+}
```
Comment (kantoniak) [inline android/app/src/main/java/labs/pooh/eaterslab/client/apis/CafeteriaApi.kt]: I'm against keeping generated files in the repository. Script you prepared ensures anyone can generate client easily.

File: android/app/build.gradle [hunk 8]
```diff
@@ -68,6 +68,23 @@ dependencies {
...
 
+
+    /**
+     * Base of data to visualisation class that gaves the ability
+     * to print plots (for example as a set of rectangles or lines)
+     */
+    implementation 'io.data2viz.d2v:scale:0.8.0-RC10'
+    implementation 'io.data2viz.d2v:core:0.8.0-RC10'
+    implementation 'io.data2viz.d2v:viz:0.8.0-RC10'
+    implementation 'io.data2viz.d2v:axis:0.8.0-RC10'
```
Comment (kantoniak) [inline android/app/build.gradle]: I seems to me that we're not using data2viz in this PR.

File: android/api_generate_kotlin.sh (hunk 2 above)
Comment (avan1235) [reply android/api_generate_kotlin.sh]: Don't want to have odl files from api, to see errors in build when api changed. mv can overwrie but not sure if it remove the old directory (just merge directories)

File: android/app/build.gradle (hunk 8 above)
Comment (avan1235) [reply android/app/build.gradle]: will remove

File: android/app/src/main/java/labs/pooh/eaterslab/MapSearchActivity.kt (hunk 3 above)
Comment (avan1235) [reply android/app/src/main/java/labs/pooh/eaterslab/MapSearchActivity.kt]: moved to gradle.properties. for now there is common version for debug/release but canbe easily changed

File: android/gradle.properties (hunk 6 above)
Comment (avan1235) [reply android/gradle.properties]: now it's used to keep also project specific properties

File: android/app/src/main/java/labs/pooh/eaterslab/client/apis/CafeteriaApi.kt (hunk 7 above)
Comment (avan1235) [reply android/app/src/main/java/labs/pooh/eaterslab/client/apis/CafeteriaApi.kt]: removed

File: android/api_generate_kotlin.sh [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+git add app/src/main/java/labs/pooh/eaterslab/client
```
Comment (kantoniak) [inline android/api_generate_kotlin.sh]: No longer required if we don't check in the client.

File: android/gradle.properties (hunk 6 above)
Comment (kantoniak) [reply android/gradle.properties]: I was thinking of some local overrides, but docs suggest keeping JVM args public anyway: https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_configuration_properties
Feel free to resolve.

File: android/api_generate_kotlin.sh (hunk 9 above)
Comment (kantoniak) [reply android/api_generate_kotlin.sh]: Oops, I meant `git add` only.

Comment (avan1235) [reply android/api_generate_kotlin.sh]: Done

**Comment by avan1235** [comment]
ending lines no longer needed as we don't keep generated files in repo

**Pull request merged**